core/src/consensus.rs

```diff
@@ -406,9 +490,27 @@ impl Tower {
         epoch_vote_accounts: &HashMap<Pubkey, (u64, Account)>,
     ) -> SwitchForkDecision {
         let root = self.lockouts.root_slot.unwrap_or(0);
+        let mut stray_restored_ancestors = HashSet::default();
+
         self.last_voted_slot()
             .map(|last_voted_slot| {
-                let last_vote_ancestors = ancestors.get(&last_voted_slot).unwrap();
+                // NOTE: Update use_stray_restored_ancestors() as well when updating this!
+                let last_vote_ancestors = if !self.is_stray_last_vote() {
+                    ancestors.get(&last_voted_slot).unwrap()
+                } else {
+                    // Use stray restored ancestors because we couldn't derive last_vote_ancestors
+                    // from given ancestors (=bank_forks)
+                    // Also, can't just return empty ancestors because we should exclude
+                    // lockouts on stray last vote's ancestors in the lockout_intervals later
+                    // in this fn.
+                    stray_restored_ancestors = self.stray_restored_ancestors();
+                    info!(
+                        "returning restored slots {:?} because of stray last vote ({})...",
+                        stray_restored_ancestors, last_voted_slot
+                    );
+                    &stray_restored_ancestors
+                };
+
                 let switch_slot_ancestors = ancestors.get(&switch_slot).unwrap();
 
                 if switch_slot == last_voted_slot || switch_slot_ancestors.contains(&last_voted_slot) {
```

Comment:
I think this check needs an extra check to account for a situation as follows:

```

1->2->3 (last_vote) -> 4 -> 5 (starting snapshot) -> 6

```
In this case, if we want to vote on `6`, even though `switch_slot_ancestors` will only include `5`,  we don't need a switching proof.

I think an  check where `|| (last_vote_slot < root && self.stray_restored_slots.is_empty())` should return, should be sufficient. We don't care about cases where `last_vote_slot > root`, because if `last_vote_slot` was an ancestor of `switch_slot`,  then `last_vote_slot` must have been replayed before `switch_slot`, and thus must exist in `BankForks` and also in `switch_slot_ancestors`.

Reply:
That's really good discussion! I'm still learning as I address your reviews.

That's great you can spot these concerns by code reviews! And I think your concern is valid. After thinking about this for a while, I think there is no need to change the code, though. In short, I believe your concern has already been addressed by the other code.

Specifically, there should be no possibility for both `self.last_voted_slot() == Some(N)` and `self.stray_restored_slots.is_empty() == true` to occur at the same time.

In other words, if and only if `self.stray_restored_slots.is_empty()`, then `self.last_voted_slot() == None` and we return `NoSwitch` from `check_switch_threshold()` by `.unwrap_or(NoSwitch)` at the very end of the `fn`'s body. Conversely, if `self.last_voted_slot() == Some(N)`, `self.stray_restored_slots.is_empty()` would always be `false` and the suggested conditional would never evaluate to `true` to enter into the special-cased `if` body anyway.

The trick is that: _The other code_ ([`do_adjust_lockouts_after_replay`](https://github.com/solana-labs/solana/pull/10718/files#r469026016)) deliberately resets `self.last_vote` if `self.lockouts.votes.is_empty()` is `true` here.

My design relationale here is that _don't let the restored tower introduce another state for tower. Minimize to violate the already-existing tower invariants. Always pretend to have been there for a while as if there was no restoration._ It's already quite stateful and super important. So, let's minimize it for controlled human's cognitive load.

That reasoning leads to this: If there were no persisted tower, there would only be 2 combinations out of 4 for the two binary-state variables: `(a) actively voting: non-empty lockouts and Some(last_vote)`, `(b) just initialized: empty lockouts and None(last_vote)`. Considering this observation, any tower state which is fully older than `root` (= no stray votes) can be regarded as being technically equivalent to the `(b)` state.

Also, normalizing fully-replayed tower to the initial/(`Default`) state of tower should be philosophically sound, forgoing to introduce _restored_ tower. I think this should be desired unless ultimately unavoidable.

The downside is that this might lead to misconception as you just were worried... For that I can extract the preceding logic into its own `fn` as a cosmetic arrangement.

To recap, once `stray_restored_slots` is assigned to `last_vote_ancestors`, it's functionally and strictly `last_vote_ancestors` in every aspect. So, there should be no leaky abstraction here, like for special-case `if`. So, the cosmetic arrangement is to indicate that intention (hopefully more clearly).

Put another way, I just tried my best to avoid to make existing conditional any longer. ;)

Reply:
@ryoqun that makes sense, I missed the subtlety with  `do_adjust_lockouts_after_replay` there.

Thanks for the well-reasoned response, I agree this approach with the default last_vote is much cleaner :)